### PR TITLE
Use bidirectional comparability in narrowing

### LINKED ISSUE #50527: [regression 4.8] type narrowed too far; valid cases considered unreachable
# Bug Report

### 🔎 Search Terms

type narrowing assertNever never unreachable 

### 🕗 Version & Regression Information

- This changed between versions 4.7.4 and 4.8.x

### ⏯ Playground Link

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/KYOwrgtgBACgTgewA7DgFwJ4BUMqgbwCgooBlNOASxAHMoBeKAcgGcLqamAaYqAOUgAjVA2bgIwuN14BVEAGsQCAO4hRTMAqWrpAX0KFqaVADMAhgGNgsRCnQYAPDjzAAHsZAATFjeSpMzsBcUABqZgA2YNZuHt5QmooqaowJ2iAAfAS8mCgAXFCBANy8AG4RUflhkcDF+oQ51uRUtPB+9qKtdpgOnf7YuMAAdE0cwWzNNOnFDfxCqL3tjAvdy-0ogwISqMHiklP1A1ByiaqrHbZ9PRf2gYPHacGpSfsHeKsA8oIAVsAWaAy8AA+ZHYLWumCBsy2cFWkPuSVhr2sH2+vzQ7zgAFEIEhMOc2phPj8-lBgfgoA18qsilAytV8ppPMATNRgJ4oLpioYQMY4OYrFBMZ5KGgHCEsFAYqA4ijif8luCMES0ZkiCRKcw2SKmMUSEhFZUsLUDEZTJZrABBTyeMUSqVeHyytEY7G4jD4rpK1F-F04zCq7IDfJMMzWnW8fUEjCG42EEyaP6UBBqMwsFj+PjAEqoAAUIHyICzqAAlAWi3AsuqABaIZRQQt1zFwRBwHNMJQoJjF2PxkCJ5NQTwIchgEwmHOWNBJ-OC4X-YFWzzFytQSgmKAT-sgQaRz2DOlRBiMcDhcLLtUkKAAeivg4QUBYCAgwDQVY4vF0kvC6dX683U+THdFUGGZ6DA3xPVueFVHPXgSBvO8HyfF831oL90w-dDrAveDb1fSgfHCVlVx8OBgEsKszEEcIgigQQwH+CwnyQSgaJ8ZQRSrBAGMlZsEDgABCOCoFTdN0EzbNW0nacgKjbsP0IOohxHMccwvDUmC1NBpD1A0V3VIMIL6W4RloHhL1pcpgGDKtgFPBBgmUfjwk8JgFN0YsgA)

### 💻 Code

```ts
enum PropertyType {
  String = 'string',
  Number = 'number',
  Unknown = 'unknown',
}

interface Property<Type extends PropertyType, Value extends unknown = unknown> {
  type: Type;
  value: Value;
}

type StringProperty = Property<PropertyType.String, string>;
type NumberProperty = Property<PropertyType.Number, number>;
type UnknownProperty = Property<PropertyType.Unknown, unknown>;

type PropertyObject =
  | StringProperty
  | NumberProperty
  | UnknownProperty

type PropertyObjectOrEmpty = PropertyObject | { type: PropertyType; value: undefined };

interface Edit<VT extends PropertyObject = PropertyObject> {
  type: 'edit';
  property: VT;
}

interface Add<VT extends PropertyObjectOrEmpty = PropertyObjectOrEmpty> {
  type: 'add';
  property: VT;
}

function assertNever(n: never): never {
  throw new Error('nope');
}

function doStuff(action: Edit | Add) {
  if (action.property.value == null) {
    // do something
  } else if (action.property.type === PropertyType.Unknown) {
    // do something else
  } else {
    // this line is reachable, but compiles without error!
    assertNever(action.property);
  }
}

doStuff({
  type: 'edit',
  property: {
    type: PropertyType.String,
    value: 'hello, world'
  }
})
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
var PropertyType;
(function (PropertyType) {
    PropertyType["String"] = "string";
    PropertyType["Number"] = "number";
    PropertyType["Unknown"] = "unknown";
})(PropertyType || (PropertyType = {}));
function assertNever(n) {
    throw new Error('nope');
}
function doStuff(action) {
    if (action.property.value == null) {
        // do something
    }
    else if (action.property.type === PropertyType.Unknown) {
        // do something else
    }
    else {
        // this line is reachable, but compiles without error!
        assertNever(action.property);
    }
}
doStuff({
    type: 'edit',
    property: {
        type: PropertyType.String,
        value: 'hello, world'
    }
});

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>
      
### 🙁 Actual behavior

Code compiles without error, but `assertNever` is hit at runtime.

### 🙂 Expected behavior

Compilation error: the type of `action.property` at the point it's passed to `assertNever` is not `never`. (4.7 does this.)

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #50527

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot pack this
@typescript-bot perf test this faster

**Comment by andrewbranch** [comment]
@ahejlsberg I think there must be other possible fixes since this isn’t directly related to #49119 which caused the regression, but this is for one of the bugs you were going to look at today

File: tests/cases/compiler/50527.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// @strict: true
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply tests/cases/compiler/50527.ts:1]: Do we though? 😛

**Review by ahejlsberg — APPROVED** [review]

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick this to release-4.8

**Pull request merged**